File: leetcode/java/LEET310.java
import java.util.*;
public class LEET310 {
    public static void main(String[] args) {
        System.out.println(findMinHeightTrees(4, new int[][] {{1,0},{1,2},{1,3}}));
    }
    public static List<Integer> findMinHeightTrees(int n, int[][] edges) {
        List<Integer> leaves = new ArrayList<>();
        if(n <= 1){
            leaves.add(0);
            return leaves;
        }
        Map<Integer, ArrayList<Integer>> graph = new HashMap<>();
        for(int[] edge : edges){
            if(!graph.containsKey(edge[0])){
                graph.put(edge[0], new ArrayList<>());
            }
            if(!graph.containsKey(edge[1])){
                graph.put(edge[1], new ArrayList<>());
            }
            graph.get(edge[0]).add(edge[1]);
            graph.get(edge[1]).add(edge[0]);
        }

        for(int i=0;i<n;i++){
            if(graph.get(i).size() == 1){
                leaves.add(i);
            }
        }

        while(n > 2){
            n -= leaves.size();
            List<Integer> newLeaves = new ArrayList<>();
            for(int leaf : leaves){
                int neighbor = graph.get(leaf).remove(0);
                graph.get(neighbor).remove(Integer.valueOf(leaf));

                if(graph.get(neighbor).size() == 1){
                    newLeaves.add(neighbor);
                }
            }

            leaves = newLeaves;
        }
        return leaves;
    }
}

File: leetcode/python/LEET310.py
from typing import *
import collections

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def findMinHeightTrees(n: int, edges: List[List[int]]) -> List[int]:
    if n <= 1:
        return [0]
    graph = collections.defaultdict(list)
    for u, v in edges:
        graph[u].append(v)
        graph[v].append(u)

    leaves = [i  for i in range(n) if len(graph[i]) == 1]

    while n > 2:
        n -= len(leaves)
        new_leaves = []
        for leaf in leaves:
            neighbor = graph[leaf].pop()
            graph[neighbor].remove(leaf)

            if len(graph[neighbor]) == 1:
                new_leaves.append(neighbor)
        
        leaves = new_leaves

    return leaves
res = findMinHeightTrees(n = 4, edges = [[1,0],[1,2],[1,3]])
print(res)